### PR TITLE
Threshold flow/v7

### PR DESCRIPTION
**Pull request by victorjulien**
SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1788

Replaces https://github.com/OISF/suricata/pull/10632
- rebase
- ticket link for minor todo (https://redmine.openinfosecfoundation.org/issues/6855)
- add support for detection_filter as well

https://redmine.openinfosecfoundation.org/issues/6822

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 20169

**Comment by catenacyber** [comment]
What is the relationship with #10925 ?

They look like conflicting at least for src/detect-engine-threshold.c

So, which one should get first ?

**Comment by victorjulien** [comment]
> What is the relationship with #10925 ?
> 
> They look like conflicting at least for src/detect-engine-threshold.c
> 
> So, which one should get first ?

Doesn't matter, I'll have to rebase the other one first. But will wait for a proper review regardless.

**Comment by catenacyber** [comment]
Ok, will pick one for review next time

**Comment by victorjulien** [comment]
Btw I see no conflict here or locally.

**Comment by catenacyber** [comment]
I get a conflict when merging both PRs

**Comment by victorjulien** [comment]
> I get a conflict when merging both PRs

They are independent, I'll deconflict them after one gets merged.

**Review by catenacyber — CHANGES_REQUESTED** [review]
Thanks for the work :-)

This PR looks good to me, but the SV one needs a small update

- CI : 🔴 I guess the SV PR needs a rebase
- Code : Looking now
- Commits segmentation : nice
- Commit messages : ok for me
- Git ID set : looks fine for me
- CLA : :-p
- Doc update : ok
- Redmine ticket : ok
- Rustfmt : not needed
- Tests : 🟠 needs rebase and left some comments there
- Dependencies added: none

File: src/detect-engine-register.h [hunk 1]
```diff
@@ -86,11 +83,26 @@ enum DetectKeywordId {
     DETECT_BYTETEST,
     DETECT_BYTEJUMP,
     DETECT_BYTEMATH,
+    DETECT_BYTE_EXTRACT,
+    DETECT_DATASET,
+    DETECT_DATAREP,
+    DETECT_BASE64_DECODE,
+    DETECT_BASE64_DATA,
+    DETECT_BSIZE,
+    DETECT_ASN1,
```
Comment (catenacyber) [inline src/detect-engine-register.h:92]: Why is ASN1 here ?

File: src/util-threshold-config.c [hunk 2]
```diff
@@ -67,11 +67,15 @@ static FILE *g_ut_threshold_fp = NULL;
...
 #define DETECT_THRESHOLD_REGEX                                                                     \
-    "^,\\s*type\\s*(limit|both|threshold)\\s*,\\s*track\\s*(by_dst|by_src|by_both|by_rule)\\s*,"   \
+    "^,\\s*type\\s*(limit|both|threshold)\\s*,\\s*track\\s*(by_dst|by_src|by_both|by_rule|by_"     \
+    "flow)\\s*,"                                                                                   \
     "\\s*count\\s*(\\d+)\\s*,\\s*seconds\\s*(\\d+)\\s*$"
 
 /* TODO: "apply_to" */
-#define DETECT_RATE_REGEX "^,\\s*track\\s*(by_dst|by_src|by_both|by_rule)\\s*,\\s*count\\s*(\\d+)\\s*,\\s*seconds\\s*(\\d+)\\s*,\\s*new_action\\s*(alert|drop|pass|log|sdrop|reject)\\s*,\\s*timeout\\s*(\\d+)\\s*$"
+#define DETECT_RATE_REGEX                                                                          \
+    "^,\\s*track\\s*(by_dst|by_src|by_both|by_rule|by_flow)\\s*,\\s*count\\s*(\\d+)\\s*,\\s*"      \
```
Comment (catenacyber) [inline src/util-threshold-config.c:76]: Should we rustify this ?

File: src/detect-engine-register.h (hunk 1 above)
Comment (victorjulien) [reply src/detect-engine-register.h:92]: it's a content inspect keyword, like the rest of the list

File: src/util-threshold-config.c (hunk 2 above)
Comment (victorjulien) [reply src/util-threshold-config.c:76]: maybe, but not now

File: src/detect-engine-threshold.c [hunk 3]
```diff
@@ -241,6 +242,70 @@ static DetectThresholdEntry *ThresholdHostLookupEntry(Host *h,
     return e;
 }
 
+typedef struct FlowVarThreshold_ {
+    uint8_t type;
+    uint8_t pad[3];
+    uint32_t idx;
```
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/detect-engine-threshold.c:248]: not used, will replace it by more padding to make the struct follow the GenericVar header format

**Comment by victorjulien** [comment]
replaced by #11029

File: src/detect-engine-threshold.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-engine-threshold.c:248]: It is used once, but it is never set

File: src/detect-engine-register.h (hunk 1 above)
Comment (catenacyber) [reply src/detect-engine-register.h:92]: Ok, I see `git grep DETECT_ASN src/detect-engine-content-inspection.c`

**Pull request closed without merging**